### ISSUE TITLE
How do I specify a license for deb packaging?

### ISSUE THREAD

**Issue by toolforger** [FibreFoX/javafx-gradle-plugin#32]
I am seeing this:

```
:jfxNative
Running [/usr/lib/jvm/java-8-openjdk-amd64/jre/bin/java, -version]
Creating app bundle: /home/jo/Projekte/javafx-tutorials/smokegraph-fx/build/jfx/native/smokegraph
Running [/usr/lib/jvm/java-8-openjdk-amd64/jre/bin/java, -version]
Running [dpkg-deb, --version]
Debian packages should specify a license.  The absence of a license will cause some linux distributions to complain about the quality of the application.
```

License is specified by placing a `copyright` file in the same directory as the  `control`, `preinstall`, `postinstall`, `prerm`, `postrm`, menu icon, and menu shortcut descriptor files, but none of the directories I tried would get copied through by `jfxNative` and the Oracle tools to where `dpkg-deb` would pick it up.

**Comment by FibreFoX** [comment]
Hi there,

you need to provide some `bundleArguments`-entry with the name `licenseFile`, this file needs to be placed inside a different folder via `additionalAppResources`-property. I will provide some example for this, as it is a bit unintuitive. It won't be in the next minutes, more in the next 8 hours (I'll do it after work).

**Comment by FibreFoX** [comment]
@toolforger 
[The bundler responsible for deb-files](https://github.com/teamfx/openjfx-8u-dev-rt/blob/e4e24b9ad7eb7536ffb2618a502d1cfe8bfbd8a3/modules/fxpackager/src/main/java/com/oracle/tools/packager/linux/LinuxDebBundler.java#L171) searches inside the provided application resources for a file which is registered inside some provided `bundleArguments`-entry called "LICENSE_FILE", which needs to be provided as `licenseFile`:

```groovy
jfx {
    // ... other stuff ...
    bundleArguments = [
        licenseFile: 'your-license.txt'
    ]
}
```

That `your-license.txt` can be provided via a separated folder and used via `additionalAppResources`-property, I'm unsure if it would be sufficient having it in `src/main/deploy` ... might work, not sure.

Hopefully this helps a bit until I have created some example-project.

**Comment by FibreFoX** [comment]
@toolforger please see [the example project](https://github.com/FibreFoX/javafx-gradle-plugin/tree/d950bed4bc6bd1edc24315c4866a6af8b7cd0939/examples/debian-installer-with-license)

Does this work for you?

**Comment by toolforger** [comment]
I've been pretty busy last week, and will be pretty busy this week.
I should be able to find time for this until next weekend though.

**Comment by FibreFoX** [comment]
@toolforger Thanks for the feedback, good luck at work :) I will do more testing on that feature in different scenarios on my virtual machines, maybe I will find something to fix.

Issue will stay open until you found some time, no pressure from me!

**Comment by FibreFoX** [comment]
Currently it is required to have the license-file being part of the `additionalAppResources`, this might change after having #38 being fixed.

**Comment by toolforger** [comment]
Yep, those resources are used by the desktop, not by application code, so they don't belong in `src/main/resources` in the first place.
There might be use cases for reusing some of that data, e.g. the icon, though that's probably less useful than one might expect because of required conversions (.bmp for Java, .ico/.png/.svg/.whatever for for Windows/Linux/Mac desktop).

**Comment by FibreFoX** [comment]
With "by the desktop", do you mean the installer? All files inside `additionalAppResources` are put into the generated application bundle, this is useful when having external-generated README-files or licenses, which (as you pointed out) do NOT belong in `src/main/resources` ;)

Is your question answered then?

**Comment by toolforger** [comment]
I misthought - it's indeed installer. Got confused because my train of thought started with those resources that the installer copies over to the desktop configuration.

**Comment by toolforger** [comment]
I just confirmed that the license file is now properly copied over.
Something still isn't quite right because the resulting .deb still displays with "License: Non-free", but that may very well be a different problem; I'll do a new issue once I get around to investigating this.
This could be quite a while into the future or never happen: I'll look into injecting jfxrt.jar into the launch4j toolchain, in the hopes that that will be less effort than investigating and working around all the rough edges and limitations of the Oracle toolchain.

**Comment by FibreFoX** [comment]
To speficy the licensetype, please take a look at [the exampe-project](https://github.com/FibreFoX/javafx-gradle-plugin/tree/master/examples/debian-installer-with-license)

> You need to specify some license-type aswell for having the `.deb`-builder and `.rpm`-builder working
correct, please make sure to use some known string for the license-type. The Fedora-project has a list
online to take from: https://fedoraproject.org/wiki/Licensing:Main?rd=Licensing#Good_Licenses


```groovy
jfx {
    // ...
    additionalAppResources = 'src/main/additionalFiles'
    bundleArguments = [
        licenseType: 'ASL 2.0',
        licenseFile: 'LICENSE'
    ]
    // ...
}
```

**Comment by toolforger** [comment]
A licenseType of CC0 is specified; don't know what's happening behind the scenes or which actor is responsible for this behaviour; all I know is that I'm not seeing any license-related warnings or errors.
Given my current priorities, I'll simply leave this uninvestigated. I mentioned this only so you know that future users might get into similar problems.

**Comment by FibreFoX** [comment]
Maybe it's something cached? I've looked into the file where the license-type gets reflected: [template.spec](https://github.com/teamfx/openjfx-8u-dev-rt/blob/e7f9dac62fe0e22116ea5b23d8f6ff5c8fd977d4/modules/fxpackager/src/main/resources/com/oracle/tools/packager/linux/template.spec)
```
Summary: APPLICATION_SUMMARY
Name: APPLICATION_PACKAGE
Version: APPLICATION_VERSION
Release: 1
License: APPLICATION_LICENSE_TYPE
Vendor: APPLICATION_VENDOR
Prefix: /opt
Provides: APPLICATION_PACKAGE
Requires: ld-linux.so.2 libX11.so.6 libXext.so.6 libXi.so.6 libXrender.so.1 libXtst.so.6 libasound.so.2 libc.so.6 libdl.so.2 libgcc_s.so.1 libm.so.6 libpthread.so.0 libthread_db.so.1
Autoprov: 0
Autoreq: 0
```

I've checked the list on the fedora project, CC0 is listed ... don't know if the version of the `dpgk-deb`-tool has something to do with this.

**Comment by toolforger** [comment]
Caching is unlikely, I opened the .deb directly.
No real surprise that CC0 is listed. (Note to readers: CC0 would be wrong to use for licensing if you bundle the JRE - Oracle's lawyers might have you for breakfast if you do this.)

I'm not seeing any `spec` file in the .deb, maybe that's the problem.
Here's the relevant snippet from my `gradle.build`:

```
jfx {
    ...
    bundleArguments = [
        licenseType: 'CC0',
        licenseFile: 'license.txt'
    ]
    ...
```

The `license.txt` file is included allright.

**Comment by FibreFoX** [comment]
Hm ... I will look into it later. That spec-file is used for generating the deb-file, so it is not required so be inside ;)

**Comment by FibreFoX** [comment]
@toolforger how do you look into the deb file to have your "non-free"-information? Generated DEB-File seems to not contain that license-type anywhere ... maybe this is only for RPM-packages?!?

**Comment by toolforger** [comment]
.debs are normal archive files (tgz format I believe), so 7zip etc. should be able to look into it (being on Ubuntu, I used Archive Manager).
Debian archives contain the license information in a machine-readable format. I noticed something was amiss when Software Center told me that it had a non-free license. Looking into the .deb confirmed that while the free-format license.txt I had provided was there, the CC0 info I had provided as licenseType was missing. My understanding is that there should be a file named `spec` that contains this information (plus maybe a few more lines).
Interestingly enough, `dpkg-deb` stopped complaining. I don't know whether that's because the license.txt was there, or because I had specified a licenseType, and I didn't go back and experiment further.

**Comment by FibreFoX** [comment]
Seems like a known bug of Ubuntu: https://bugs.launchpad.net/ubuntu/+source/software-center/+bug/435183

As far as I understand this, anything that does NOT come from Ubuntu-repos, is marked as non-free. Even the linked debian spec for the license lists CC0: https://www.debian.org/doc/packaging-manuals/copyright-format/1.0/#license-specification

I think this can be closed then, as all your information are reflected inside the deb-package now.
